Refactor module internals and enhance clarity

LaureanoOlocco opened the pull request:
## 📝 **Description:**

This PR includes seven major changes to improve compatibility and feature coverage in Ghidralation:

---

### **1. Refactor `set_name` and `get_name`:**

* **Unified program context:** Replaced `currentProgram` references with `cp.currentProgram` for consistent context handling across scripts and tests.
* **Correct address handling:** Removed redundant `minAddress` addition in `get_name` to prevent incorrect lookups.
* **Namespace resolution:** Updated to use `program.getNamespaceManager().getGlobalNamespace()`, replacing the deprecated `symbol_table.getGlobalNamespace()`.
* **Improved symbol creation:** Ensured proper handling of `SN_FORCE` (unique naming) and `SN_NODUMMY`.
* **Test validation:** Confirmed that `test_basic_symbol_creation` passes and that `get_name` correctly retrieves newly created symbols.

---

### **2. Implement `idc.op_stroff`:**

* Added an implementation for `idc.op_stroff` to emulate IDA's structure offset operand formatting:

  * Retrieves the instruction at a given address.
  * Resolves structures by name or `Structure` object.
  * Applies operand comments to visually represent "struct offset" in disassembly.
  * Supports operand index selection (`0`, `1`, or `-1` for all operands).
* Added `test/test_op_stroff.py`:

  * Validates module import.
  * Finds the first valid instruction automatically.
  * Creates a dummy `TEST_STRUCT` and applies it to an operand.
  * Verifies that the operand annotation was added successfully.

---

### **3. Implement `SetType`:**

* Implemented `idc.SetType()` equivalent for assigning or clearing C-style types at specific addresses.
* Supports:

  * Parsing type declarations using Ghidra's `DataTypeParser`.
  * Clearing existing data when an empty type string is provided.
  * Safe application via `DataUtilities.createData`.
* Improvements:

  * Replaced magic number `0` with constant `USE_DEFAULT_LENGTH` for clarity.
  * Corrected misuse of `FlatProgramAPI` in `createData()` call — now properly uses `cp.currentProgram`.
  * Added structured docstring and exception handling for robustness.
* Future-ready for type integration with structure-aware plugins.

---

### **4. Add `SEGATTR_START` support to `get_segm_attr`:**

* **Complete attribute coverage:** Extended `get_segm_attr()` to support both `SEGATTR_START` (segment begin address) and `SEGATTR_END` (segment end address), matching IDA's API specification.
* **Implementation:** Added conditional handling for `SEGATTR_START` that returns `block.getStart().getOffset() - min_address`.
* **API consistency:** Updated function docstring to reflect support for both attributes.
* **Compatibility:** Ensures scripts using `SEGATTR_START` (commonly used in IDA plugins) now work correctly.

---

### **5. Extract hardcoded segment name strings to constants:**

* **Centralized constants:** Replaced hardcoded strings `"EXTERNAL"` and `"extern"` in `get_segm_name()` with module-level constants:
  * `EXTERNAL_SEGMENT_NAME = "EXTERNAL"`
  * `EXTERN_SEGMENT_ALIAS = "extern"`
* **Benefits:**
  * Prevents typos and inconsistencies across the codebase.
  * Simplifies maintenance when segment naming conventions change.
  * Improves code readability with descriptive constant names.
  * Enables reuse across multiple functions.

---

### **6. Implement platform-agnostic `import_type`:**

* **Problem:** Previous implementation was UEFI-specific, returning hardcoded IDs for EFI types (`EFI_GUID`, `EFI_SYSTEM_TABLE`, etc.), violating the platform-agnostic design principle.
* **Solution:** Completely reimplemented to match IDA's `import_type()` behavior:
  * Searches for types in Ghidra's `DataTypeManager` (current program's types).
  * Falls back to `BuiltInDataTypeManager` for built-in types (primitives, common structures).
  * Uses `dtm.resolve()` with `DataTypeConflictHandler.DEFAULT_HANDLER` to properly import types with dependency resolution.
  * Returns `UniversalID.hashCode()` as the type ID (or name hash as fallback).
  * Returns `BADNODE` constant on failure, matching IDA convention.
* **New imports added:**
  * `from ghidra.app.util.datatype import DataTypeConflictHandler`
  * `from ghidra.program.model.data import BuiltInDataTypeManager`
* **New constant:** `BADNODE = -1` (IDA's standard error value for type/node operations).
* **Impact:** Now works with **any platform** (ELF, PE, Mach-O, etc.), not just UEFI binaries.

---

### **7. Implement platform-agnostic `get_struc_id`:**

* **Problem:** Like `import_type`, this was UEFI-specific, returning hardcoded values for `EFI_GUID` and `EFI_BOOT_SERVICES`.
* **Solution:** Reimplemented to dynamically look up structure IDs:
  * Searches the current program's `DataTypeManager` for the structure by name.
  * Uses `findDataTypes()` as fallback if exact path match fails.
  * Validates that the found type is actually a `Structure` (not an enum, union, etc.).
  * Returns `UniversalID.hashCode()` as structure ID.
  * Returns `BADNODE` on failure (structure not found or invalid type).
* **Impact:** Generic implementation that works with structures from any binary format or custom types.

---

## **Why:**

These changes address multiple issues:

* `set_name/get_name` previously failed due to incorrect offset adjustments and deprecated namespace APIs.
* Scripts lacked support for `op_stroff`, breaking compatibility with IDAPython plugins that rely on it.
* `SetType` was missing, blocking plugins that manipulate types (e.g., for structs or custom variables).
* **`get_segm_attr`** was incomplete, missing `SEGATTR_START` support that many IDA scripts rely on.
* **Hardcoded strings** in `get_segm_name()` increased maintenance burden and error risk.
* **`import_type` and `get_struc_id`** were platform-specific (UEFI-only), breaking the intended design of `idc.py` as a **platform-agnostic translation layer**. According to code review feedback, platform-specific logic belongs in plugins, not in core API wrappers.
* The original implementations of `import_type` and `get_struc_id` didn't actually import types from libraries—they just returned magic numbers, making them unusable for real type management workflows.

This refactor standardizes context handling, updates APIs for modern Ghidra versions, introduces missing API coverage, restores platform independence to the core translation layer, and implements correct IDA API behavior for type management functions.

---

## **Impact:**

* Symbols (`set_name/get_name`) now behave correctly in GUI and headless modes.
* `op_stroff` allows porting IDA scripts involving structure operand annotations.
* `SetType` enables type application workflows required by many IDAPython plugins.
* **`get_segm_attr`:** Full compatibility with IDA scripts using `SEGATTR_START` to get segment boundaries.
* Segment name constants: Cleaner, more maintainable code with centralized string definitions.
* **`import_type`:** Now actually imports types from Ghidra's type libraries (built-ins, archives, program types) instead of returning hardcoded values. Works with **ELF, PE, Mach-O, and any custom structures**, not just UEFI.
* **`get_struc_id`:** Dynamically retrieves structure IDs for any structure in the program, enabling proper structure-aware IDA plugin compatibility.
* Error handling: Proper `BADNODE` returns on failure, with diagnostic messages for debugging.
* Improved maintainability and test coverage for both symbol and operand handling.


---

## **Testing Instructions:**

1. **Run symbol tests:**

   ```bash
   ./run_test.sh
   ```

## **References:**

* [IDA Python Documentation - idc module](https://python.docs.hex-rays.com/idc/index.html)
* [idc.get_segm_attr()](https://python.docs.hex-rays.com/idc/index.html#idc.get_segm_attr)
* [idc.import_type()](https://python.docs.hex-rays.com/idc/index.html#idc.import_type)
* [idc.get_struc_id()](https://python.docs.hex-rays.com/idc/index.html#idc.get_struc_id)

lautalom committed:
Fix some things

lautalom committed:
GCL

lautalom committed:
findcrypt replay

lautalom committed:
syms2elf refactor

lautalom committed:
delete author thing

Commit:
ida_bytes

Commit:
ida_bytes correct

Commit:
ida_bytes correct

Commit:
ida_kernwin

Commit:
idaapi

Commit:
ida_nalt

Commit:
ida_name

Commit:
ida_name correct

Commit:
ida_ua

lautalom committed:
fix addressing? findcrypt

Commit:
idautils test

lautalom committed:
fix funcs syms2elf

Commit:
idc

Commit:
idc idautils ida_bytes

Commit:
invoke

Commit:
invoke

lautalom committed:
invoke

lautalom committed:
invoke

Commit:
invoke

Commit:
refactor 1

lautalom committed:
idaapi

Commit:
idaapi

Commit:
idautils

Commit:
ida_bytes

Commit:
idautils

Commit:
idaapi

Commit:
idaapi

Commit:
latest stage

LaureanoOlocco committed:
Fix README.md

LaureanoOlocco committed:
Fix README

LaureanoOlocco committed:
Fix README.md

LaureanoOlocco committed:
Fix README

LaureanoOlocco committed:
Fix README

LaureanoOlocco committed:
Fix README

LaureanoOlocco committed:
refactor: reorganize project structure into src/ directory

- Move all Python modules (ida_*, idc.py, idaapi.py, etc.) into src/ folder
- Maintain same functionality but improve project organization
- Separate source code from root-level files like README.md and images/
- This change prepares the project for better modularity and testing structure

LaureanoOlocco committed:
Add set_name func

LaureanoOlocco committed:
Add .gitignore to ignore __pycache__ and other files

LaureanoOlocco committed:
feat: add set_name implementation

LaureanoOlocco committed:
Add PyGhidra headless execution script and reorganize test files

LaureanoOlocco committed:
fix: unify cp.currentProgram usage and address handling

- Use cp.currentProgram consistently.
- Remove redundant minAddress in get_name.
- Use NamespaceManager for global namespace.
- Ensure correct symbol creation and lookup.

LaureanoOlocco committed:
Merge remote-tracking branch 'origin/ghidralation' into dev

LaureanoOlocco committed:
Fix comments

LaureanoOlocco committed:
Fix comments

LaureanoOlocco committed:
feat(idc): implement op_stroff and add test suite

dgutson commented on src/idc.py:
```diff
@@ -7,9 +6,13 @@
-from array import array
 from ghidra.program.flatapi import FlatProgramAPI
+from ghidra.program.model.data import DataUtilities
+from ghidra.program.model.data import DataTypePath  # Fixed import
```
no need for comment

dgutson commented on src/idc.py:
```diff
@@ -35,8 +39,12 @@ def get_segm_attr(segea, attr):
     minAddress = cp.currentProgram.minAddress.getOffset()
     fcp = FlatProgramAPI(cp.currentProgram)
     if attr == SEGATTR_END:
```
else? what does it return?

dgutson commented on src/idc.py:
```diff
@@ -107,24 +118,32 @@ def import_type(idx, type_name):
-        fcp.toAddr(ea+minAddress)).getStart().getOffset()
-    return block-minAddress
+    block = (
+        cp.currentProgram.getMemory()
```
```python
cp.currentProgram.getMemory()
 .getBlock(fcp.toAddr(ea + minAddress))
```

repeated many times; move to a function with a meaningful name (what is that?)

dgutson commented on src/idc.py:
```diff
@@ -154,11 +173,9 @@ def get_operand_value(ea, n):
         value = op.reg
-    elif op.type & OperandType.SCALAR and not\
-            (op.type & OperandType.ADDRESS):
+    elif op.type & OperandType.SCALAR and not (op.type & OperandType.ADDRESS):
```
beware of stylish changes. Use yapf or black, try to preserve format.

dgutson commented on src/idc.py:
```diff
@@ -167,7 +184,7 @@ def get_operand_value(ea, n):
 
-    if value == 0x3c or value == 0x60:
+    if value == 0x3C or value == 0x60:
         print("HERE")
```
???

dgutson commented on src/idc.py:
```diff
@@ -167,7 +184,7 @@ def get_operand_value(ea, n):
     else:
         value = -1
 
-    if value == 0x3c or value == 0x60:
```
what are these values? No magic numbers. Use constants with meaningful names.

dgutson commented on src/idc.py:
```diff
@@ -188,11 +205,13 @@ def get_struc_id(struc):
         dtm = currentProgram.getDataTypeManager()
-        selectionDialog = DataTypeSelectionDialog(tool, dtm, -1, AllowedDataTypes.FIXED_LENGTH)
+        selectionDialog = DataTypeSelectionDialog(
+            tool, dtm, -1, AllowedDataTypes.FIXED_LENGTH
```
pls use named args (for example, I have no clue what the -1 is)

LaureanoOlocco committed:
refactor: apply review feedback to SetType and segment helpers

- Replaced GUI type selection with direct parser (SetType)
- Used named arguments for clarity (DataTypeSelectionDialog)
- Removed magic numbers and added placeholder constants
- Deduplicated getBlock calls into reusable helper
- Preserved formatting with black

LaureanoOlocco committed:
Fix main.py for search plugins in plugins/

LaureanoOlocco committed:
Add plugins directory to gitignore

LaureanoOlocco committed:
Removed magic number in SetType(), improved function docstrings.

LaureanoOlocco committed:
Removed unnecessary __all__ variable

dgutson commented on src/ida_diskio.py:
```diff
@@ -1,19 +1,18 @@
-        else str(idir[0]) + path_to_version
-    )
+    path_to_version = os.path.join("IDAPro","Python", "7xx")
+    path = str(idir[1]) if len(idir) else str(idir[0]) 
```
"idir" is very confusing in this context: IDA dir? Installation Dir? (I think the latter). Please rename.

dgutson replied:
what are 0 and 1? No magic numbers, define local constants and use them.

dgutson commented on src/ida_diskio.py:
```diff
@@ -1,19 +1,18 @@
+    path_to_version = os.path.join("IDAPro","Python", "7xx")
+    path = str(idir[1]) if len(idir) else str(idir[0]) 
+    path += path_to_version
+    if not os.path.isdir(path):
```
just `os.makedirs('/your/full/path', exist_ok=True)`

dgutson commented on src/ida_kernwin.py:
```diff
@@ -0,0 +1,31 @@
+def ask_form(*args):
+    pass
+
+def ask_str(*args):
```
this function does nothing (?). I think it should be `pass` as the others.

dgutson commented on src/ida_name.py:
```diff
@@ -0,0 +1,239 @@
+GN_NOT_ISRET = 0x0100  # For dummy names: do not use return location
+
+
+def get_name(ea, flags):
```
why flags are not used? Maybe take a look at set_name to see whether anyone is to be handled?

If you can confirm that flags are not needed here, document it with a comment.

LaureanoOlocco replied:
According to the latest documentation version, ida_name.get_name now accepts only one parameter. [get_name doc](https://python.docs.hex-rays.com/ida_name/index.html#ida_name.get_name)

dgutson commented on src/ida_name.py:
```diff
@@ -0,0 +1,239 @@
+    return set_name(ea, name, SN_FORCE | SN_NOCHECK)
+
+
+# __all__ = [
```
remove commented code.

dgutson commented on src/ida_ua.py:
```diff
@@ -0,0 +1,72 @@
+from ghidra.program.model.lang import OperandType
+import cp
+
+REGS = {'RAX': 0, 'EAX': 0, 'AX': 0,
```
DANGER: This is x86-specific in a generic file. Submit an issue to yourself to refactor this to place this code in arch-specific files, and address other archs as well.

dgutson commented on src/idaapi.py:
```diff
@@ -16,68 +17,88 @@
+PLUGIN_MOD = 1
+PLUGIN_PROC = 64
+PLUGIN_FIX = 128
+BADADDR = hex(4294967295)
```
???

This is not well implemented.
Consider defining the EA type (effective address) as (and, BTW start adding types)
```python
class BADADDR:
    pass

EA = int | BADADDR
```

Also, there is another definition of BADADDR as -1 (which may or may not numerically match). Unify the definitions, decide where to store the types and constants in terms of who uses them. Design the layering and "X-uses-Y" relationships without cyclic imports.

dgutson commented on src/ida_name.py:
```diff
@@ -0,0 +1,239 @@
+from ghidra.program.model.address import Address
+
+# SN_ Constants (set_name() flags)
+SN_CHECK = 0x00  # Fail if the name contains invalid characters.
```
Consider using a Flag or IntFlag class for these definitions.

dgutson commented on src/ida_name.py:
```diff
@@ -0,0 +1,239 @@
+)
+
+# GN_ Constants (get_name() flags)
+GN_VISIBLE = 0x0001  # Replace forbidden characters by SUBSTCHAR
```
likewise above.

By using a (e.g. Flag) subtype, you ensure that flags are not wrongly combiner (e.g. OR-ed).

dgutson commented on src/idaapi.py:
```diff
@@ -16,68 +17,88 @@
+get_imagebase = ida_nalt.get_imagebase
+
 
 def action_desc_t(*args):
```
Not sure if thes args come from Ghidra, but I would create a class to wrap and provide methods to manipulate/operate them.

I'm seeing a lot of global functions receiving *args as argument. I would encapsulate all of them in a class. However, if these functions (like this one) are part of the API that you are mimicking, then ignore this comment.

dgutson commented on src/idc.py:
```diff
@@ -0,0 +1,399 @@
+    Get segment attribute value by EA.
+
+    @param ea: any address in the segment (IDA-style EA)
+    @param attr: attribute (currently only SEGATTR_END supported)
```
no need for SEGATTR_BEGIN? Does it exist?

LaureanoOlocco replied:
Yes it exists. SEGATTR_START. More information [here.](https://python.docs.hex-rays.com/8.5/d1/dcc/namespaceidc.html#ab4e433403eec1ebab022d362877d8cc6) And the documentation of the function [here](https://python.docs.hex-rays.com/8.5/d1/dcc/namespaceidc.html#a9c7e951a659f70c58aff3285270a9d43)

dgutson commented on src/idc.py:
```diff
@@ -0,0 +1,399 @@
+        return ""
+
+    name = block.getName()
+    if name == "EXTERNAL":
```
Move this to a constant somewhere.

dgutson commented on src/idc.py:
```diff
@@ -0,0 +1,399 @@
+    pass
+
+
+def import_type(idx, type_name):
```
DANGER: this function is platform (UEFI)-specific, therefore should be in a plugin, not in the translation layer that should be platform agnostic (e.g. ELF, PE for windows, etc.).

Try to remove it from here and place it in a plugin or, if not possible for now, in a temporary platform-specific file.

dgutson commented on src/idc.py:
```diff
@@ -0,0 +1,399 @@
+    return ida_bytes.prev_head(ea, maxea)
+
+
+def get_struc_id(struc):
```
Again, platform-specific.

dgutson commented on src/idc.py:
```diff
@@ -0,0 +1,399 @@
+    pass
+
+
+def set_name(ea, name, flags=0):
```
Why I am seeing a redefinition of these functions? ida_name.py, ida_api, and here.

This is the most important finding of this review: revise the design, there's something very wrong with this.

LaureanoOlocco replied:
This is how it is in the documentation. They have the same name but they are for different purposes.

<figure style="text-align:center;">
  <img src="https://github.com/user-attachments/assets/d8fb5ec4-0d9f-4457-802c-d04641eb9db3" width="737" height="349" alt="Figure 1 - idc.set_name">
  <figcaption>Figure 1 - idc.set_name</figcaption>
</figure>

---

<figure style="margin:0 auto; width:fit-content;">
  <img src="https://github.com/user-attachments/assets/7db0f7a1-9cfe-409f-af57-2444a7a42242"
       width="743" height="868" alt="Figure 2 - ida_name.set_name"
       style="display:block; margin:0 auto;">
  <figcaption style="text-align:center;">Figure 2 - ida_name.set_name</figcaption>
</figure>

dgutson replied:
Ok, it is because they are different tools (ida vs idc).

Commit:
Fix run_test.sh

Signed-off-by: LaureanoOlocco <laureano.olocco@eclypsium.com>

Commit:
Fix run_test.sh

Signed-off-by: LaureanoOlocco <laureano.olocco@eclypsium.com>

LaureanoOlocco committed:
Fix run_test.sh

Signed-off-by: LaureanoOlocco <laureanoolocco@gmail.com>

LaureanoOlocco committed:
Refactor idc.py: Add SEGATTR_START, constants, and platform-agnostic type functions

- Add SEGATTR_START support to get_segm_attr() for complete segment attribute coverage
- Extract hardcoded segment names to constants (EXTERNAL_SEGMENT_NAME, EXTERN_SEGMENT_ALIAS)
- Implement platform-agnostic import_type() that imports from Ghidra's DataTypeManager
- Implement platform-agnostic get_struc_id() for dynamic structure ID lookup
- Add BADNODE constant for IDA-compatible error handling
- Remove UEFI-specific hardcoded logic in favor of generic implementations

These changes restore platform independence to idc.py, making it work with any binary format (ELF, PE, Mach-O, etc.) instead of just UEFI.

dgutson commented on src/idc.py:
```diff
@@ -19,10 +21,16 @@
 import ida_name
 
-SEGATTR_END = 8
+SEGATTR_START = 0
```
IntEnum?

LaureanoOlocco committed:
refactor(idc): replace integer constants with IntEnum classes
